### PR TITLE
Basic docs

### PR DESCRIPTION
**Pull request by TomorrowIsAnOtherDay**
add documentation for basic classes: 
- parl.Model
- parl.Algorithm
- parl.Agent

### CODE REVIEW

File: parl/core/fluid/algorithm.py [hunk 1]
```diff
@@ -23,16 +23,47 @@
...
+        model = Model()
+        dqn = parl.algorithms.DQN(model, lr=1e-3)
 
-    Before creating a customized algorithm, please do check algorithms of PARL.
-    Most common used algorithms like DQN/DDPG/PPO/A3C/IMPALA have been provided in `parl.algorithms`,
-    go and have a try.
+    Attributes:
+        model(``parl.Model``): a neural network that represents a policy or a Q-value function.
+
+    Pulich Functions:
```
Comment (zenghsh3) [inline parl/core/fluid/algorithm.py]: Publish

File: parl/core/fluid/agent.py [hunk 2]
```diff
@@ -54,71 +89,70 @@ def __init__(self, algorithm, gpu_id=None):
...
+            | This function will be called automatically in the initialization function.
         
-        To build the program, you may need to do the following:
-            a. Create a new program of fluid with program guard;
-            b. Define data input layers;
-            c. Pass the data variable defined in step b to learn/predict/sample of algorithm;
+        To build a program, you must do the following:
+            a. Create a fluid program with ``fluid.program_guard()``;
+            b. Define data layers for feeding the data;
+            c. build various programs(e.g., learn_program, predict_program) with data layers defined in step b.
```
Comment (fuyw) [inline parl/core/fluid/agent.py]: c. Build ...

File: parl/core/fluid/agent.py [hunk 3]
```diff
@@ -26,11 +26,48 @@
 
 
 class Agent(AgentBase):
+    """
+    | `alias`: ``parl.Agent``
+    | `alias`: ``parl.core.fluid.agent.Agent``
+
+    | Agent is one of three basic classes of PARL.
```
Comment (fuyw) [inline parl/core/fluid/agent.py]: one of the three ...

File: parl/core/fluid/agent.py [hunk 4]
```diff
@@ -26,11 +26,48 @@
...
+
+      .. code-block:: python
+
+        import parl
+
+        class MyAgent(parl.Agent):
+            def __init__(self, algorithm, act_dim):
+                super(MyAgent, self).__init__(algorithm)
+                self.act_dim = act_dim
+      This class will automatically initialize all the parameters of neural network, and provides an executor for users to run programs(self.fluid_executor).
```
Comment (fuyw) [inline parl/core/fluid/agent.py]: This class will initialize the neural network parameters automatically, and provides an executor for users to run the programs (self.fluid_executor).

File: parl/core/fluid/agent.py [hunk 5]
```diff
@@ -26,11 +26,48 @@
...
+
+    Attributes:
+        gpu_id (int): deprecated. specify which GPU to be used. -1 if to use the CPU.
+        fluid_executor (fluid.Executor): executor for running programs of the agent.
+        alg (parl.algorithm): algorithm of this agent.
+
+    Public Functions:
+        - ``build_program`` (**abstract function**): build various programs for the agent to interact with outer environment.
+        - ``set_weights``: returns a list containing all the parameters of self.alg.
```
Comment (fuyw) [inline parl/core/fluid/agent.py]: return ... ( for consistency )

File: parl/core/fluid/agent.py [hunk 6]
```diff
@@ -26,11 +26,48 @@
...
+        - ``get_weights``: copy parameters from ``set_weights()`` to this agent.
+        - ``sample``: return a noisy action to perform exploration according to the policy.
+        - ``predict``: return a action given current observation.
```
Comment (fuyw) [inline parl/core/fluid/agent.py]: an action

File: parl/core/fluid/algorithm.py [hunk 7]
```diff
@@ -23,16 +23,47 @@
...
+        - ``get_weights``: return a list containing all the parameters of the current model.
+        - ``set_weights``: copy parameters from ``get_weights()`` to the model.
+        - ``sample``: return a noisy action to perform exploration according to the policy.
+        - ``predict``: return a action given current observation.
```
Comment (fuyw) [inline parl/core/fluid/algorithm.py]: an action

File: parl/core/fluid/agent.py [hunk 8]
```diff
@@ -54,71 +89,70 @@ def __init__(self, algorithm, gpu_id=None):
...
-           1. Accept numpy data as input;
-           2. Feed numpy data;
-           3. Run predict program defined in `build_program`.
+        This function feeds the observation into the prediction program defined in ``build_program()``. It is often used in the evaluation stage.
         """
         raise NotImplementedError
 
     def sample(self, *args, **kwargs):
-        """Sample the action when given the observation of the enviroment.
+        """return an action with noise when given the observation of the environment.
```
Comment (fuyw) [inline parl/core/fluid/agent.py]: Return

File: parl/core/fluid/algorithm.py [hunk 9]
```diff
@@ -23,16 +23,47 @@
...
 
 class Algorithm(AlgorithmBase):
-    """Algorithm defines the way how we update the model.
-    
-    To implement a new algorithm, you may need implement the learn/predict/sample functions.
+    """
+    | `alias`: ``parl.Algorithm``
+    | `alias`: ``parl.core.fluid.algorithm.Algorithm``
+
+    | ``Algorithm`` defines the way how to update the parameters of the ``Model``. That is, we define loss functions and the optimizer of the neural network. An ``Algorithm`` has at least a model.
```
Comment (fuyw) [inline parl/core/fluid/algorithm.py]: This is where we define ...

File: parl/core/fluid/model.py [hunk 10]
```diff
@@ -24,45 +24,46 @@
 
 
 class Model(ModelBase):
-    """A `Model`, a collection of `parl.layers`, is owned by an `Algorithm`. 
+    """
+    | `alias`: ``parl.Model``
+    | `alias`: ``parl.core.fluid.agent.Model``
+
+    | ``Model`` is a base class of PARL for the neural network. A ``Model`` is usually a policy or Q-value function, and predicts an action or an estimate according to the environmental observation.
```
Comment (fuyw) [inline parl/core/fluid/model.py]: or Q-value function, which predicts

File: parl/core/fluid/model.py [hunk 11]
```diff
@@ -24,45 +24,46 @@
...
-       # Need initialize program before calling sync_weights_to
-       fluid_executor.run(fluid.default_startup_program()) 
-       ...
+    Attributes:
+        model_id(str): each model instance has its uniqe model_id.
 
-       # synchronize parameters
-       model.sync_weights_to(target_model)
+    Public Functions:
+        - ``sync_weights_to``: Synchronize parameters of the current model to another model.
```
Comment (fuyw) [inline parl/core/fluid/model.py]: synchronize ...

File: parl/core/fluid/model.py [hunk 12]
```diff
@@ -200,10 +226,10 @@ def parameters(self):
     @deprecated(
         deprecated_in='1.2', removed_in='1.3', replace_function='get_weights')
     def get_params(self):
-        """Get numpy arrays of parameters in the model.
+        """ returns a list containing all the parameters of current model.
```
Comment (fuyw) [inline parl/core/fluid/model.py]: Return a list

File: parl/core/fluid/algorithm.py (hunk 1 above)
Comment (TomorrowIsAnOtherDay) [reply parl/core/fluid/algorithm.py]: done

File: parl/core/fluid/agent.py (hunk 3 above)
Comment (TomorrowIsAnOtherDay) [reply parl/core/fluid/agent.py]: done

File: parl/core/fluid/agent.py [hunk 13]
```diff
@@ -26,11 +26,49 @@
...
+      This class will initialize the neural network parameters automatically, and provides an executor for users to run the programs (self.fluid_executor).
+
+    Attributes:
+        gpu_id (int): deprecated. specify which GPU to be used. -1 if to use the CPU.
+        fluid_executor (fluid.Executor): executor for running programs of the agent.
+        alg (parl.algorithm): algorithm of this agent.
+
+    Public Functions:
+        - ``build_program`` (**abstract function**): build various programs for the agent to interact with outer environment.
+        - ``set_weights``: return a list containing all the parameters of self.alg.
```
Comment (zenghsh3) [inline parl/core/fluid/agent.py]: comments of `set_weights` and `get_weights` are misplaced.
And `get_weights` will return a dict: {`model_attribute_name`: list}.

File: parl/core/fluid/algorithm.py [hunk 14]
```diff
@@ -23,16 +23,46 @@
...
+        model = Model()
+        dqn = parl.algorithms.DQN(model, lr=1e-3)
 
-    Before creating a customized algorithm, please do check algorithms of PARL.
-    Most common used algorithms like DQN/DDPG/PPO/A3C/IMPALA have been provided in `parl.algorithms`,
-    go and have a try.
+    Attributes:
+        model(``parl.Model``): a neural network that represents a policy or a Q-value function.
+
+    Pulish Functions:
```
Comment (zenghsh3) [inline parl/core/fluid/algorithm.py]: Public

File: parl/core/fluid/algorithm.py [hunk 15]
```diff
@@ -23,16 +23,46 @@
...
+        - ``get_weights``: return a list containing all the parameters of the current model.
```
Comment (zenghsh3) [inline parl/core/fluid/algorithm.py]: get_weights will return a dict: {model_attribute_name: list}.

**Review by zenghsh3 — CHANGES_REQUESTED** [review]

File: parl/core/fluid/algorithm.py (hunk 14 above)
Comment (TomorrowIsAnOtherDay) [reply parl/core/fluid/algorithm.py]: done

File: parl/core/fluid/algorithm.py (hunk 15 above)
Comment (TomorrowIsAnOtherDay) [reply parl/core/fluid/algorithm.py]: done

**Review by zenghsh3 — DISMISSED** [review]
LGTM

**Review by zenghsh3 — APPROVED** [review]

**Pull request merged**